### PR TITLE
Support VR test including TestStream for Spark runner in streaming mode

### LINKED ISSUE #22472: [Improvement]: Support TestStream in SparkRunner to validate streaming runner
### What would you like to happen?

Currently the Spark streaming runner has its own set of limited validations.
It would be good to leverage the validates runner "framework" for this as well, which requires support for TestStream/

### Issue Priority

Priority: 2

### Issue Component

Component: runner-spark

### PR DESCRIPTION
**Pull request by mosche**
Run VR tests for Spark streaming runner rather than custom tests (test are already run as part of the "normal" unit test run).

If `forceStreaming` is set to `true`, the `TestSparkRunner` will replace `Read.Bounded` with `UnboundedReadFromBoundedSource` so tests are run in streaming mode.
Additionally this PR adds support for `TestStream`.

Closes #22472

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by mosche** [comment]
Run Spark ValidatesRunner

**Comment by mosche** [comment]
Unfortunately I'm stuck with some flaky tests. It looks like watermarks are not advanced in a deterministic way.
Below some logs of `org.apache.beam.sdk.schemas.AvroSchemaTest.testAvroPipelineGroupBy` (edited for readability).

Successful run (watermark advanced early enough, so that timer is triggered):

```
14:42:52,938 [3] TRACE SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: input elements: [ValueInGlobalWindow{value=Row1, pane=NO_FIRING}]
14:42:52,940 [3] TRACE SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: non expired input elements: [ValueInGlobalWindow{value=Row1, pane=NO_FIRING}]
14:42:52,949 [3] TRACE WindowTracing  - ReduceFnRunner.scheduleGarbageCollectionTimer: Scheduling at GLOBALW_MAX for key:Row2; window:GlobalWindow where inputWatermark:BOUNDEDW_MIN; outputWatermark:null
14:42:52,957 [3] TRACE WindowTracing  - WatermarkHold.addHolds: element hold at GLOBALW_MAX is on time for key:Row2; window:GlobalWindow; inputWatermark:BOUNDEDW_MIN; outputWatermark:null
14:42:52,960 [3] DEBUG SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: timerInternals before advance are SparkTimerInternals{highWatermark=BOUNDEDW_MIN, synchronizedProcessingTime=EPOCH, timers=[TimerData{timerId=0, timerFamilyId=, namespace=Window(GlobalWindow), timestamp=GLOBALW_MAX, outputTimestamp=GLOBALW_MAX, domain=EVENT_TIME, deleted=false}], inputWatermark=BOUNDEDW_MIN}
14:42:52,961 [3] DEBUG SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: timers eligible for processing are [] [inputWatermark: BOUNDEDW_MIN]
14:42:52,962 [3] TRACE SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: output elements are  0
```
```
14:42:53,137 [spark-listener-group-appStatus] INFO  GlobalWatermarkHolder  - Put new watermark block: {0=SparkWatermarks{lowWatermark=BOUNDEDW_MIN, highWatermark=BOUNDEDW_MAX, synchronizedProcessingTime=NOW}}
```
```
14:42:53,146 [15] DEBUG SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: timerInternals before advance are SparkTimerInternals{highWatermark=BOUNDEDW_MAX, synchronizedProcessingTime=NOW, timers=[TimerData{timerId=0, timerFamilyId=, namespace=Window(GlobalWindow), timestamp=GLOBALW_MAX, outputTimestamp=GLOBALW_MAX, domain=EVENT_TIME, deleted=false}], inputWatermark=BOUNDEDW_MIN}
14:42:53,146 [15] DEBUG SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: timers eligible for processing are [TimerData{timerId=0, timerFamilyId=, namespace=Window(GlobalWindow), timestamp=GLOBALW_MAX, outputTimestamp=GLOBALW_MAX, domain=EVENT_TIME, deleted=false}] [inputWatermark: BOUNDEDW_MAX]
```
```
14:42:53,146 [15] DEBUG WindowTracing  - ReduceFnRunner: Received timer key:Row2; window:GlobalWindow; data:TimerData{timerId=0, timerFamilyId=, namespace=Window(GlobalWindow), timestamp=GLOBALW_MAX, outputTimestamp=GLOBALW_MAX, domain=EVENT_TIME, deleted=false} with inputWatermark:BOUNDEDW_MAX; outputWatermark:null
14:42:53,148 [15] DEBUG WindowTracing  - ReduceFnRunner: Cleaning up for key:Row2; window:GlobalWindow with inputWatermark:BOUNDEDW_MAX; outputWatermark:null
14:42:53,148 [15] DEBUG WindowTracing  - WatermarkHold.extractAndRelease: for key:Row2; window:GlobalWindow; inputWatermark:BOUNDEDW_MAX; outputWatermark:null
14:42:53,149 [15] DEBUG WindowTracing  - WatermarkHold.extractAndRelease.read: clearing for key:Row2; window:GlobalWindow
14:42:53,150 [15] DEBUG WindowTracing  - describePane: ON_TIME pane (prev was null) for key:Row2; windowMaxTimestamp:GLOBALW_MAX; inputWatermark:BOUNDEDW_MAX; outputWatermark:null; isLateForOutput:false
14:42:53,152 [15] TRACE WindowTracing  - ReduceFnRunner.onTrigger: outputWindowedValue key:Row2 value:[Row1] at GLOBALW_MAX
14:42:53,152 [15] DEBUG WindowTracing  - WatermarkHold.clearHolds: For key:Row2; window:GlobalWindow; inputWatermark:BOUNDEDW_MAX; outputWatermark:null
14:42:53,153 [15] TRACE SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: output elements are TimestampedValueInGlobalWindow{value=KV{Row2, [Row1]}, timestamp=GLOBALW_MAX, pane=PaneInfo{isFirst=true, isLast=true, timing=ON_TIME, index=0, onTimeIndex=0}} 1
```

Failed run (watermark is advanced too late, element is lost):
```
14:41:51,453 [3] TRACE SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: input elements: [ValueInGlobalWindow{value=Row1, pane=NO_FIRING}]
14:41:51,455 [3] TRACE SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: non expired input elements: [ValueInGlobalWindow{value=Row1, pane=NO_FIRING}]
14:41:51,463 [3] TRACE WindowTracing  - ReduceFnRunner.scheduleGarbageCollectionTimer: Scheduling at GLOBALW_MAX for key:Row2; window:GlobalWindow where inputWatermark:BOUNDEDW_MIN; outputWatermark:null
14:41:51,471 [3] TRACE WindowTracing  - WatermarkHold.addHolds: element hold at GLOBALW_MAX is on time for key:Row2; window:GlobalWindow; inputWatermark:BOUNDEDW_MIN; outputWatermark:null
14:41:51,474 [3] DEBUG SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: timerInternals before advance are SparkTimerInternals{highWatermark=BOUNDEDW_MIN, synchronizedProcessingTime=EPOCH, timers=[TimerData{timerId=0, timerFamilyId=, namespace=Window(GlobalWindow), timestamp=GLOBALW_MAX, outputTimestamp=GLOBALW_MAX, domain=EVENT_TIME, deleted=false}], inputWatermark=BOUNDEDW_MIN}
14:41:51,474 [3] DEBUG SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: timers eligible for processing are [] [inputWatermark: BOUNDEDW_MIN]
14:41:51,476 [3] TRACE SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: output elements are  0
```
```
14:41:51,658 [15] DEBUG SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: timerInternals before advance are SparkTimerInternals{highWatermark=BOUNDEDW_MIN, synchronizedProcessingTime=EPOCH, timers=[TimerData{timerId=0, timerFamilyId=, namespace=Window(GlobalWindow), timestamp=GLOBALW_MAX, outputTimestamp=GLOBALW_MAX, domain=EVENT_TIME, deleted=false}], inputWatermark=BOUNDEDW_MIN}
14:41:51,658 [15] DEBUG SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: timers eligible for processing are [] [inputWatermark: BOUNDEDW_MIN]
14:41:51,658 [15] TRACE SparkGroupAlsoByWindowViaWindowSet  - Group.ByFields/ToKvs/GroupByKey: output elements are  0
```
```
14:41:51,662 [spark-listener-group-appStatus] INFO  GlobalWatermarkHolder  - Put new watermark block: {0=SparkWatermarks{lowWatermark=BOUNDEDW_MIN, highWatermark=BOUNDEDW_MAX, synchronizedProcessingTime=NOW}}
```

**Comment by mosche** [comment]
Run Spark ValidatesRunner

**Comment by mosche** [comment]
Run Spark ValidatesRunner

**Comment by mosche** [comment]
Run Spark ValidatesRunner

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java [hunk 1]
```diff
@@ -105,6 +105,9 @@ public void testLateDataAccumulating() {
             .advanceWatermarkTo(instant.plus(Duration.standardMinutes(6)))
             // These elements are late but within the allowed lateness
             .addElements(TimestampedValue.of(4L, instant), TimestampedValue.of(5L, instant))
+            .advanceWatermarkTo(instant.plus(Duration.standardMinutes(10)))
```
Comment (mosche) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java:108]: @kennknowles Maybe you could answer this? I'm wondering if this is an issue of the Spark streaming runner (and how this is handled by other runners) or if it's a lack of my own understanding.

**Comment by mosche** [comment]
Run Java PreCommit

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java (hunk 1 above)
Comment (kennknowles) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java:108]: Just responding to let you know I have been on vacation and I will look at this later today.

Comment (mosche) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java:108]: Thanks a lot! I'm off as well for a bit, so no rush on this.

Comment (mosche) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java:108]: @kennknowles Finally back to this, if you could have a look it would be great:)

File: runners/spark/spark_runner.gradle [hunk 2]
```diff
@@ -260,21 +265,60 @@ def validatesRunnerStreaming = tasks.register("validatesRunnerStreaming", Test)
...
-    "--forceStreaming=true",
-    "--enableSparkMetricSinks=true",
-  ])
-  systemProperty "beamTestPipelineOptions", pipelineOptions
+  systemProperties sparkTestProperties(["--enableSparkMetricSinks": "true", "--forceStreaming": "true"])
 
   classpath = configurations.validatesRunner
-  testClassesDirs += files(project.sourceSets.test.output.classesDirs)
+  testClassesDirs += files(
+    project(":sdks:java:core").sourceSets.test.output.classesDirs,
```
Comment (kennknowles) [inline runners/spark/spark_runner.gradle:272]: This change makes sense. I don't understand how it worked before. Was it not actually running the VR tests?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java [hunk 3]
```diff
@@ -153,4 +192,84 @@ private static void awaitWatermarksOrTimeout(
     } while ((timeoutMillis -= batchDurationMillis) > 0
         && globalWatermark.isBefore(stopPipelineWatermark));
   }
+
+  /**
+   * Override factory to replace {@link Read.Unbounded} with {@link UnboundedReadFromBoundedSource}
+   * to force streaming mode.
+   */
+  private static class UnboundedReadFromBoundedSourceOverrideFactory<T>
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java:200]: This seems useful as a general thing that could be in `runners-core-construction` FWIW.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/SparkRunnerStreamingContextFactory.java [hunk 4]
```diff
@@ -85,7 +85,14 @@ public JavaStreamingContext call() throws Exception {
     EvaluationContext ctxt = new EvaluationContext(jsc, pipeline, options, jssc);
     // update cache candidates
     SparkRunner.updateCacheCandidates(pipeline, translator, ctxt);
-    pipeline.traverseTopologically(new SparkRunner.Evaluator(translator, ctxt));
+    try {
+      pipeline.traverseTopologically(new SparkRunner.Evaluator(translator, ctxt));
+    } catch (RuntimeException e) {
+      jssc.stop(false, false);
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/SparkRunnerStreamingContextFactory.java:91]: Identifiers like `jsc` and `jssc` are very unreadable to me who is not doing Spark stuff every day.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+import org.apache.spark.streaming.dstream.InputDStream;
+import org.joda.time.Instant;
+import scala.Option;
+import scala.reflect.ClassTag;
+
+public class TestDStream<T> extends InputDStream<WindowedValue<T>> {
+  private final Coder<WindowedValue<T>> coder;
+
+  @SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED") // not intended for use after serialization
+  private final transient List<TestStream.Event<T>> events;
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java:52]: I think all transient fields should be `@Nullable`. Is checkerframework disabled at the whole package level?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    super(ssc, classTag());
+    this.coder = WindowedValue.getFullCoder(test.getValueCoder(), GlobalWindow.Coder.INSTANCE);
+    this.events = test.getEvents();
+  }
+
+  @Override
+  public Option<RDD<WindowedValue<T>>> compute(Time validTime) {
+    Preconditions.checkStateNotNull(events);
+
+    TestStream.Event<T> event = insertEmptyBatch ? null : nextEvent();
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java:72]: Confusing with `this.event`. I would rename `this.event` to `this.currentEventIndex` or something.

**Review by kennknowles — COMMENTED** [review]
This is super useful! Thank you!

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java [hunk 7]
```diff
@@ -415,7 +418,12 @@ public void testEarlyPanesOfWindow() {
   }
 
   @Test
-  @Category({ValidatesRunner.class, UsesTestStream.class, UsesTestStreamWithMultipleStages.class})
+  @Category({
+    ValidatesRunner.class,
+    UsesTestStream.class,
+    UsesTestStreamWithMultipleStages.class,
+    UsesStatefulParDo.class
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java:425]: This small change should be an independent commit that comes first. Then if there is any trouble we will only revert the bigger commit but not this fixup.

**Comment by kennknowles** [comment]
Run Spark ValidatesRunner

File: runners/spark/spark_runner.gradle (hunk 2 above)
Comment (mosche) [reply runners/spark/spark_runner.gradle:272]: Yes, in fact VR test were never run/supported for Spark in streaming mode. I just stumbled on this accidentally when I started looking into bugs related to onWindowExpiration :(
Instead there was some custom tests in the module that try to mimic the VR test, but they only cover a very small part (and also run as unit tests).

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java (hunk 5 above)
Comment (mosche) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java:52]: Hmm, checker is enabled, not sure if that's a miss or due to the findbugs suppression. I'll have a look

**Comment by mosche** [comment]
@kennknowles FYI, a significant number of VR tests are constantly failing here. If I run them independently they usually succeed. It looks like there's some indeterminism around watermark propagation in the runner, see https://github.com/apache/beam/issues/23129.
Wondering, would you know anyone who's familiar with that code?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java (hunk 7 above)
Comment (mosche) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/testing/TestStreamTest.java:425]: Done ✔️ 
https://github.com/apache/beam/pull/23202

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/SparkRunnerStreamingContextFactory.java (hunk 4 above)
Comment (mosche) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/SparkRunnerStreamingContextFactory.java:91]: I created a separate issue/PR for this: https://github.com/apache/beam/pull/23204

File: runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java (hunk 3 above)
Comment (mosche) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/TestSparkRunner.java:200]: Happy to do that though I'm not entirely sure if the factory is of much value by itself. I also had to fix the outputs in a non trivial way using a visitor after the replacement, see https://github.com/apache/beam/pull/22620/files#diff-d81f49eb0330230bd03ce6cd33b5f70f59c443aac57741e877ececbada32b16bR246-R274

I couldn't find a way to achieve this in `mapOutputs` of the override factory itself.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java (hunk 5 above)
Comment (mosche) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java]: Checker is running, but not catching this apparently. I added `@Nullable`

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java (hunk 6 above)
Comment (mosche) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/translation/streaming/TestDStream.java:70]: Done ✔️

**Comment by kennknowles** [comment]
I don't know if anyone currently around would be familiar with SparkRunner watermark propagation.

**Comment by kennknowles** [comment]
I think it is valuable to get these tests running and disable them. The test and list of disabled tests can be a real representation of the current state. That way things that are green can stay green.

**Comment by kennknowles** [comment]
run spark validatesrunner

**Comment by mosche** [comment]
Run Spark ValidatesRunner

**Comment by mosche** [comment]
I took a bit of a turn here after validating my initial approach replacing bounded sources with `UnboundedReadFromBoundedSource` with VR tests in Flink:

- Tests that failed likely due to watermark issues with the Spark runner (#23129, see [test results](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_PR/494/)) ran fine with Flink suggesting there really is a major problem (in streaming mode).

- Nevertheless, it also showed that the approach is somehow flawed. Some bounded test cases simply cannot be forced into a streaming execution, e.g. any GroupByKey will fail on the GlobalWindow if there's no trigger set.

The initial reason for this approach was to prevent the Spark runner from failing when streaming was forced via pipeline options in VR tests for bounded test cases: Spark refuses to start if there's no streaming workload scheduled. 
Instead `TestSparkRunner` now just detects the translation mode and acts accordingly.

Unfortunately, this hides any watermark issues uncovered above as VR tests succeed.

**Comment by kennknowles** [comment]
> * Nevertheless, it also showed that the approach is somehow flawed. Some bounded test cases simply cannot be forced into a streaming execution, e.g. any GroupByKey will fail on the GlobalWindow if there's no trigger set.

In the Beam model, this condition is that a GroupByKey of an _unbounded_ PCollection in global window must have a trigger. But you can still have a bounded PCollection in streaming mode.

So the summary is:

 - forcing a run in streaming mode, but leaving bounded PCollections as bounded is OK
 - automatically making all PCollections unbounded is flawed (but still can be useful to find bugs sometimes)

**Comment by mosche** [comment]
@kennknowles fine to merge this?

**Comment by aromanenko-dev** [comment]
@kennknowles kind ping, are you ok to merge it?

**Review by kennknowles — APPROVED** [review]
